src/textual/command_palette.py

```diff
@@ -0,0 +1,700 @@
+    """The function to call when the command is chosen."""
+
+    command_text: str
+    """The command text associated with the hit, as plain text.
+
+    This is the text that will be placed into the `Input` field of the
+    [command palette][textual.command_palette.CommandPalette] when a
+    selection is made.
+    """
+
+    command_help: str | None = None
+    """Optional help text for the command."""
+
+    def __lt__(self, other: CommandSourceHit) -> bool:
+        return self.match_value < other.match_value
+
+    def __eq__(self, other: CommandSourceHit) -> bool:
+        return self.match_value == other.match_value
+
+
+CommandMatches: TypeAlias = AsyncIterator[CommandSourceHit]
+"""Return type for the command source match hunting method."""
+
+
+class CommandSource(ABC):
+    """Base class for command palette command sources.
+
+    To create a source of commands inherit from this class and implement
+    [`hunt_for`][textual.command_palette.CommandSource.hunt_for].
+    """
+
+    def __init__(self, screen: Screen, match_style: Style | None = None) -> None:
+        """Initialise the command source.
+
+        Args:
+            screen: A reference to the active screen.
+        """
+        self.__screen = screen
+        self.__match_style = match_style
+
+    @property
+    def focused(self) -> Widget | None:
+        """The currently-focused widget in the currently-active screen in the application.
+
+        If no widget has focus this will be `None`.
+        """
+        return self.__screen.focused
+
+    @property
+    def screen(self) -> Screen[object]:
+        """The currently-active screen in the application."""
+        return self.__screen
+
+    @property
+    def app(self) -> App[object]:
+        """A reference to the application."""
+        return self.__screen.app
+
+    @property
+    def match_style(self) -> Style | None:
+        """The preferred style to use when highlighting matching portions of the `match_display`."""
+        return self.__match_style
+
+    def matcher(self, user_input: str, case_sensitive: bool = False) -> Matcher:
+        """Create a [fuzzy matcher][textual._fuzzy.Matcher] for the given user input.
+
+        Args:
+            user_input: The text that the user has input.
+            case_sensitive: Should match be case sensitive?
+
+        Returns:
+            A [fuzzy matcher][textual._fuzzy.Matcher] object for matching against candidate hits.
+        """
+        return Matcher(
+            user_input, match_style=self.match_style, case_sensitive=case_sensitive
+        )
+
+    @abstractmethod
+    async def hunt_for(self, user_input: str) -> CommandMatches:
+        """A request to hunt for commands relevant to the given user input.
+
+        Args:
+            user_input: The user input to be matched.
+
+        Yields:
+            Instances of [`CommandSourceHit`][textual.command_palette.CommandSourceHit].
+        """
+        raise NotImplemented
+
+    def run(
```

Comment:
Does this need to be public? Feels like something that should be internal.

`run` seems an odd name, given it returns a callable but doesn't actually run it.

Reply:
Yes, it needs to be public as it's a helper method that's part of the public API for making a command source (see its docstring for the explanation of what it does). I'm open to a better name if you have one, but it felt reasonably declarative: "when this command is selected `run` this thing that might be sync or async", sort of thing.

Reply:
Now removed as per the chat this morning.